quickwit/quickwit-doc-mapper/src/doc_mapper/deser_num.rs

```diff
@@ -0,0 +1,226 @@
+                    std::any::type_name::<T>(),
+                    v
+                ))
+            })
+        }
+
+        fn visit_u64<E>(self, v: u64) -> Result<T, E>
+        where E: de::Error {
+            T::deserialize(v.into_deserializer()).map_err(|_: E| {
+                de::Error::custom(format!(
+                    "expected {}, got inconvertible JSON number `{}`",
+                    std::any::type_name::<T>(),
+                    v
+                ))
+            })
+        }
+
+        fn visit_f64<E>(self, v: f64) -> Result<T, E>
+        where E: de::Error {
+            T::deserialize(v.into_deserializer()).map_err(|_: E| {
+                de::Error::custom(format!(
+                    "expected {}, got inconvertible JSON number `{}`",
+                    std::any::type_name::<T>(),
+                    v
+                ))
+            })
+        }
+
+        fn visit_map<M>(self, mut map: M) -> Result<T, M::Error>
+        where M: de::MapAccess<'de> {
+            let json_value: Value =
+                Deserialize::deserialize(de::value::MapAccessDeserializer::new(&mut map))?;
+            Err(de::Error::custom(error_message(json_value, self.coerce)))
+        }
+
+        fn visit_seq<S>(self, mut seq: S) -> Result<T, S::Error>
+        where S: de::SeqAccess<'de> {
+            let json_value: Value =
+                Deserialize::deserialize(de::value::SeqAccessDeserializer::new(&mut seq))?;
+            Err(de::Error::custom(error_message(json_value, self.coerce)))
+        }
+
+        fn visit_none<E>(self) -> Result<Self::Value, E>
+        where E: de::Error {
+            Err(de::Error::custom(error_message("null", self.coerce)))
+        }
+
+        fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+        where E: de::Error {
+            Err(de::Error::custom(error_message(v, self.coerce)))
+        }
+    }
+
+    deserializer
+        .deserialize_any(CoerceVisitor {
+            coerce,
+            marker: std::marker::PhantomData,
+        })
+        .map_err(|err| err.to_string())
+}
+
+fn error_message<T: Display>(got: T, coerce: bool) -> String {
+    if coerce {
+        format!("expected JSON number or string, got `{}`", got)
+    } else {
+        format!("expected JSON, got `{}`", got)
+    }
+}
+
+pub fn deserialize_i64<'de, D>(deserializer: D, coerce: bool) -> Result<i64, String>
+where D: Deserializer<'de> {
+    deserialize_num_with_coerce(deserializer, coerce)
+}
+
+pub fn deserialize_u64<'de, D>(deserializer: D, coerce: bool) -> Result<u64, String>
+where D: Deserializer<'de> {
+    deserialize_num_with_coerce(deserializer, coerce)
+}
+
+pub fn deserialize_f64<'de, D>(deserializer: D, coerce: bool) -> Result<f64, String>
+where D: Deserializer<'de> {
+    deserialize_num_with_coerce(deserializer, coerce)
+}
+
+#[cfg(test)]
+mod tests {
+    use serde_json::json;
+
+    use super::*;
+
+    #[test]
```

Comment:
we need to test what happens for stuff that is out of bound.

- "-1" to u64
- u64::MAX to i64
- 1.2 to u64
- "1.2" to u64